### PR TITLE
doc: coding guidelines: Add a new rule for macro name collisions

### PR DESCRIPTION
**Pull request by carlescufi**
This commit is the outcome of the discussion that has taken place in multiple forums:

Discord:
https://discord.com/channels/720317445772017664/1014241011989487716/1032623375228608522

GitHub:
https://github.com/zephyrproject-rtos/zephyr/issues/51371 https://github.com/zephyrproject-rtos/zephyr/pull/50239

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

### CODE REVIEW

File: doc/contribute/coding_guidelines/index.rst
Comment (stephanosio) [inline doc/contribute/coding_guidelines/index.rst:1274]: another common pattern I have seen is to `#undef` and re-define, which is something that should certainly be avoided.

```suggestion
other implementations. In particular, they must not be prefixed to place them in
a Zephyr-specific namespace, re-defined using `#undef`, or conditionally
excluded from compilation using ``#ifndef``.
```

**Review by de-nordic — DISMISSED** [review]
Finally carved in stone. Thanks!

File: doc/contribute/coding_guidelines/index.rst
Comment (stephanosio) [inline doc/contribute/coding_guidelines/index.rst:1287]: typo

```suggestion
Since, in the context of this documentation, Zephyr is considered the central or
main project, it should implement the non-namespaced versions of the
macros. Given that Zephyr uses a fork of the corresponding upstream for each
```

File: doc/contribute/coding_guidelines/index.rst
Comment (cfriedt) [inline doc/contribute/coding_guidelines/index.rst:1272]: > must not be prefixed..

This is ultra-confusing because we already have a large number of `Z_` prefixed macros in our codebase.

What exactly is wrong with changing non-prefixed, zephyr-specific macros to have a zephyr-specific prefix (i.e. a kind of namespace like `Z_` which we use extensively already)?

I don't mean undefining them or guarding them with ifndef.

E.g. since ARRAY_SIZE is used everywhere else in the universe, why not define a Z_ARRAY_SIZE that the Zephyr project uses everywhere to future-proof it?

**Review by cfriedt — DISMISSED** [review]
Either the wording is very confusing or this rule is being broken already

File: doc/contribute/coding_guidelines/index.rst
Comment (stephanosio) [reply doc/contribute/coding_guidelines/index.rst:1272]: `Z_` (or `z_` for that matter) is supposed to be for Zephyr internal functions.

Comment (cfriedt) [reply doc/contribute/coding_guidelines/index.rst:1272]: Sure - than another prefix if not that.

**Review by henrikbrixandersen — DISMISSED** [review]

File: doc/contribute/coding_guidelines/index.rst
Comment (cfriedt) [reply doc/contribute/coding_guidelines/index.rst:1272]: @carlescufi - can answer the question above for me?

I missed the dev meeting, but just going over the minutes quickly, I've found zero notes justifying why we *should not* "namespace ourselves" (to use the terminology from the minutes). This PR simply says "they must not be prefixed to place them in a Zephyr-specific namespace".

It seems rather non-democratic to try to codify this as policy with no justification other than some people's personal preference.

```
hal: nxp: ARRAY_SIZE collision [Carles]
Issue: https://github.com/zephyrproject-rtos/zephyr/issues/51371
Some HAL modules have their own implementation of common macros like ARRAY_SIZE, MIN, MAX
Three options:
Require modules to namespace
Namespace ourselves
Make HAL use zephyr macros instead
Greg: this is an example of the global namespace problem. There is no way we can force namespaces on downstream projects. Mcuboot is trying integrate with zephyr and with mynewt
David: making HAL use zephyr macros was rejected by the NXP SDK team. We need a process to invoke a wider scope CI run to catch these conflicts
Carles: two modules could collide with each other
Bjarki: we have a lot of defines, none of which are namespaced
David: these macros have been in HAL for years, but we didn’t discover the issue until recently when the include order changed
Kumar: fine for HAL to have the definition and use internally, but poor form to expose externally. User probably expects macros to be defined by OS, not by a HAL
Anas: ifndef violates our coding guidelines. Need header hygiene in the HALs
Carles: agree with Kumar that it is counterintuitive for the HAL to define these common macros instead of the OS. When building for zephyr, the NXP could align to zephyr standards
David: if you want to force users to get Zephyr’s version of these definitions, then you include the zephyr header first
Carles: don’t think it’s future proof to require users to include the zephyr header first. Should implement option 1 or 3 in module
Greg: zephyr isn’t the center of the constellation
Anas: there are at least 15 modules in zephyr today that use ARRAY_SIZE
Carles will send a doc PR
```

I would highly suggest we re-evaluate this.

Comment (gmarull) [reply doc/contribute/coding_guidelines/index.rst:1272]: I think that if a prefix is chosen, it should apply to all Zephyr APIs, not just _common macros_. I'd personally have done that from the very beginning, and treat Zephyr as just another library, not the center of the world. But I think it's too late for a change like that, so IMO it's better to just give Zephyr certain naming privileges.

Comment (carlescufi) [reply doc/contribute/coding_guidelines/index.rst:1272]: > What exactly is wrong with changing non-prefixed, zephyr-specific macros to have a zephyr-specific prefix (i.e. a kind of namespace like Z_ which we use extensively already)?

This was discussed extensively in a couple of meetings and the consensus was to leave the Zephyr ones not namespaced, and namespace the rest.

**Review by stephanosio — DISMISSED** [review]

File: doc/contribute/coding_guidelines/index.rst
Comment (carlescufi) [reply doc/contribute/coding_guidelines/index.rst:1272]: @cfriedt there is no "justification" other than this was the general consensus. The purpose of this PR is to collect objections as well, so feel free to object and, if necessary, we will take it to the TSC for a vote. The info about what was discussed can be found in the different links I posted in the PR description.

Comment (nashif) [reply doc/contribute/coding_guidelines/index.rst:1272]: > @cfriedt there is no "justification" other than this was the general consensus.

The consensus is based on a few justifications:
1. prefixing and namespacing every macro we have is rather disruptive.
2. Even if it is decided to prefix one or two macros, where do we stop?
3. Why does everyone get to use MIN/MAX/ARRAY_SIZE but zephyr?
4. The issue itself is mostly related to header hygene and inclusion of the wrong headers in the wrong places, so prefixing is just a workaround
5. imagine zephyr with a prefix ZEPHYR_MACRO, two modules used to build an app, both define MACRO and both are exposing them publicly, what happens then? Still did not solve the problem.
6. What prefix would chose? Z_ is for private, ZEPHYR_MIN and ZEPHYR_MAX seems rather weird and long, we will spend the next year bikeshedding over a prefix...
7. and there is more....

But as Carles said, this is not "decided", this is how we bring the discussion from the meeting to the wider community. So, objections and other views are welcome.

Comment (cfriedt) [reply doc/contribute/coding_guidelines/index.rst:1272]: > > @cfriedt there is no "justification" other than this was the general consensus.
> 
> The consensus is based on a few justifications:
> 
> 1. prefixing and namespacing every macro we have is rather disruptive.

Except that it doesn't have to be at all, as I mentioned nearly 2 months ago [here](https://github.com/zephyrproject-rtos/zephyr/pull/50239#issuecomment-1256738002).

The default can be to *enable* defining `ARRAY_SIZE`, `MIN`, and `MAX` to do exactly what they do now. 

e.g.
```
#ifdef CONFIG_DEFINE_ARRAY_SIZE
#define ARRAY_SIZE(x) Z_ARRAY_SIZE(x)
#endif
```

> 2. Even if it is decided to prefix one or two macros, where do we stop?

That's a "slippery slope" logical fallacy. Currently we're talking about 3 macros. Let's stop at 3 for now. If another macro becomes promblematic in the future, then it's 4. Not really a problem.

> 3. Why does everyone get to use MIN/MAX/ARRAY_SIZE but zephyr?

I think this is the root of the issue - it's inconvenient. People don't want to change their habits. That's understandable.

> 4. The issue itself is mostly related to header hygene and inclusion of the wrong headers in the wrong places, so prefixing is just a workaround

Incorrect - prefixing (i.e. namespacing) is to explicitly avoid collisions and do so in a way that is sustainable. It's the same reason we prefix virtually all public kernel API with `K_` or `k_`. 

Simply do not guard the definition with `#ifndef` and then we don't have include order problems, but give end-users the option (via Kconfig) to opt-out.

> 5. imagine zephyr with a prefix ZEPHYR_MACRO, two modules used to build an app, both define MACRO and both are exposing them publicly, what happens then? Still did not solve the problem.
> 6. What prefix would chose? Z_ is for private, ZEPHYR_MIN and ZEPHYR_MAX seems rather weird and long, we will spend the next year bikeshedding over a prefix...
> 7. and there is more....

**Review by de-nordic — DISMISSED** [review]

**Review by gmarull — DISMISSED** [review]

File: doc/contribute/coding_guidelines/index.rst
Comment (cfriedt) [reply doc/contribute/coding_guidelines/index.rst:1272]: I missed the dev meeting, and sorry for coming off sharp. I personally don't care about this too much.. I just wanted to highlight that there is a better way of doing things technically, but I won't block this anymore. No need to vote on it in the TSC meeting.

Comment (nashif) [reply doc/contribute/coding_guidelines/index.rst:1272]: > Is "must" really the right word to use? Maybe "should"?

yeah, maybe should, or if this  " must be implemented, and its implementation verified.  Period! " then *shall*. 

And btw, we need to be consistent with this language across our docs, should, shall and must, *shall* have the same meaning across all zephyr docs, or it will be confusing.

**Review by tbursztyka — DISMISSED** [review]

**Comment by stephanosio** [comment]
@carlescufi Shall we merge this now that nobody objects to it ?

**Review by jhedberg — DISMISSED** [review]

**Comment by carlescufi** [comment]
Note that my intentions was for this to apply to all past, present and future common macro names, not the ones that already exist in Zephyr. This would be in line with PRs like [this one](https://github.com/zephyrproject-rtos/zephyr/pull/52488).
However, this is not settled and requires discussion in the TSC.

**Comment by nashif** [comment]
> Note that my intention was for this to apply to all past, present and future common macro names, not only the ones that already exist in Zephyr. This would be in line with PRs like https://github.com/zephyrproject-rtos/zephyr/pull/52488.

I do not think the language in this PR does address new macros being introduced into Zephyr. When we decide to use a new macro or renaming an existing macro to something that conflicts with modules, we are asking modules to change something they used already and did not conflict with zephyr before. I guess the question is, what is "common" and does the fact that Linux uses the macro makes it common? 
i have an issue with renaming or prefixing a macro we use in zephyr, common or not, so it does not conflict with someone else using the same macro. But I do not understand why we have to use a Linux macro name when we can  continue using the existing name, or if we do not like it, use some other name that does not conflict with anyone. 

We might be able to "negotiate" and have modules, especially HALs, move to what ever zephyr is using or rename, but this is calling for trouble and conflicts in general when we decide to adopt something into the tree that pre-existed elsewhere. 

We need to be cautious with the language of the rule if we decide to adopt this approach.

**Comment by gmarull** [comment]
> > Note that my intention was for this to apply to all past, present and future common macro names, not only the ones that already exist in Zephyr. This would be in line with PRs like #52488.
> 
> I do not think the language in this PR does address new macros being introduced into Zephyr. When we decide to use a new macro or renaming an existing macro to something that conflicts with modules, we are asking modules to change something they used already and did not conflict with zephyr before. I guess the question is, what is "common" and does the fact that Linux uses the macro makes it common? i have an issue with renaming or prefixing a macro we use in zephyr, common or not, so it does not conflict with someone else using the same macro. But I do not understand why we have to use a Linux macro name when we can continue using the existing name, or if we do not like it, use some other name that does not conflict with anyone.
> 
> We might be able to "negotiate" and have modules, especially HALs, move to what ever zephyr is using or rename, but this is calling for trouble and conflicts in general when we decide to adopt something into the tree that pre-existed elsewhere.
> 
> We need to be cautious with the language of the rule if we decide to adopt this approach.

Existing modules can always include Zephyr headers and keep using the same macro names, as soon as the module is happy with Zephyr implementation. If we give Zephyr certain naming privileges, they should apply to any case; otherwise, we're basically privileging modules just because they came before this rule, and makes things inconsistent.

**Review by jfischer-no — DISMISSED** [review]

**Review by henrikbrixandersen — DISMISSED** [review]

**Comment by carlescufi** [comment]
@nashif @gmarull I've modified the text so that it covers additional ground, and I've tried to clarify it a bit.

File: doc/contribute/coding_guidelines/index.rst
Comment (gmarull) [inline doc/contribute/coding_guidelines/index.rst:1281]:
```suggestion
implementation in Zephyr should use that name. While there is a subjective and
```

File: doc/contribute/coding_guidelines/index.rst
Comment (gmarull) [inline doc/contribute/coding_guidelines/index.rst:1289]:
```suggestion
the form of modules. Those modules are typically independent projects that may
```

**Review by henrikbrixandersen — DISMISSED** [review]

File: doc/contribute/coding_guidelines/index.rst
Comment (mbolivar-nordic) [inline doc/contribute/coding_guidelines/index.rst:1282]:
```suggestion
non-measurable component to what "commonly used" means, the ultimate goal is
```

**Comment by mbolivar-nordic** [comment]
@carlescufi one thing that doesn't seem to be addressed here is what happens if the collision is between two modules.

I am seeing this here: https://github.com/zephyrproject-rtos/zephyr/pull/53502

Example:

```
/__w/zephyr/modules/fs/fatfs/include/ffconf.h:295: error: "STRINGIZE" redefined [-Werror]
  295 | #define STRINGIZE(x) STRINGIZE_AGAIN(x)
      | 
In file included from /__w/zephyr/modules/hal/silabs/gecko/emlib/inc/em_common.h:34,
                 from /__w/zephyr/modules/hal/silabs/gecko/emlib/inc/em_core.h:34,
                 from /__w/zephyr/modules/hal/silabs/gecko/emlib/inc/em_bus.h:34,
                 from ../../../../../../../soc/arm/silabs_exx32/efm32gg11b/soc.h:25,
                 from ../../../../../../../include/zephyr/arch/arm/aarch32/cortex_m/cmsis.h:19,
                 from ../../../../../../../include/zephyr/arch/arm/aarch32/mpu/arm_mpu_v7m.h:11,
                 from ../../../../../../../include/zephyr/arch/arm/aarch32/mpu/arm_mpu.h:14,
                 from ../../../../../../../include/zephyr/arch/arm/aarch32/arch.h:210,
                 from ../../../../../../../include/zephyr/arch/cpu.h:19,
                 from ../../../../../../../include/zephyr/kernel_includes.h:33,
                 from ../../../../../../../include/zephyr/kernel.h:17,
                 from /__w/zephyr/zephyr/subsys/fs/fat_fs.c:9:
/__w/zephyr/modules/hal/silabs/gecko/common/inc/sl_common.h:82: note: this is the location of the previous definition
   82 | #define STRINGIZE(X) #X
```

Does anyone have an idea about what we should do in this case?

**Comment by gmarull** [comment]
> @carlescufi one thing that doesn't seem to be addressed here is what happens if the collision is between two modules.
> 
> I am seeing this here: #53502
> 
> Example:
> 
> ```
> /__w/zephyr/modules/fs/fatfs/include/ffconf.h:295: error: "STRINGIZE" redefined [-Werror]
>   295 | #define STRINGIZE(x) STRINGIZE_AGAIN(x)
>       | 
> In file included from /__w/zephyr/modules/hal/silabs/gecko/emlib/inc/em_common.h:34,
>                  from /__w/zephyr/modules/hal/silabs/gecko/emlib/inc/em_core.h:34,
>                  from /__w/zephyr/modules/hal/silabs/gecko/emlib/inc/em_bus.h:34,
>                  from ../../../../../../../soc/arm/silabs_exx32/efm32gg11b/soc.h:25,
>                  from ../../../../../../../include/zephyr/arch/arm/aarch32/cortex_m/cmsis.h:19,
>                  from ../../../../../../../include/zephyr/arch/arm/aarch32/mpu/arm_mpu_v7m.h:11,
>                  from ../../../../../../../include/zephyr/arch/arm/aarch32/mpu/arm_mpu.h:14,
>                  from ../../../../../../../include/zephyr/arch/arm/aarch32/arch.h:210,
>                  from ../../../../../../../include/zephyr/arch/cpu.h:19,
>                  from ../../../../../../../include/zephyr/kernel_includes.h:33,
>                  from ../../../../../../../include/zephyr/kernel.h:17,
>                  from /__w/zephyr/zephyr/subsys/fs/fat_fs.c:9:
> /__w/zephyr/modules/hal/silabs/gecko/common/inc/sl_common.h:82: note: this is the location of the previous definition
>    82 | #define STRINGIZE(X) #X
> ```
> 
> Does anyone have an idea about what we should do in this case?

I guess patch both modules to either use namespaced macro (ideal if accepted by upstream maintainers), or, make them use Zephyr macros.

**Comment by LingaoM** [comment]
Not only that, we should specify a specification for all global labels (variables, functions, etc.). see https://github.com/zephyrproject-rtos/zephyr/pull/50581

**Comment by carlescufi** [comment]
> > @carlescufi one thing that doesn't seem to be addressed here is what happens if the collision is between two modules.
> > I am seeing this here: #53502
[snip]
> > Does anyone have an idea about what we should do in this case?
> 
> I guess patch both modules to either use namespaced macro (ideal if accepted by upstream maintainers), or, make them use Zephyr macros.

@mbolivar-nordic @gmarull I've added a sentence at the end to cover this edge case.

File: doc/contribute/coding_guidelines/index.rst
Comment (nashif) [inline doc/contribute/coding_guidelines/index.rst:1283]: replace
`offer the Zephyr user macro names that they may already be familiar with`
with
`offer users familiar macros`
?

Comment (nashif) [reply doc/contribute/coding_guidelines/index.rst:1283]: > If a macro name is commonly used in several other well-known open source projects...

What if the 'well-known open source project' is a module of zephyr?

Comment (sjg20) [reply doc/contribute/coding_guidelines/index.rst:1283]: Before accepting a module into Zephyr we should check that the module can use Zephyr's macros or has names that don't conflict.

**Review by sjg20 — DISMISSED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Review by gmarull — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]

**Review by LingaoM — APPROVED** [review]

**Comment by carlescufi** [comment]
Needs final ACK in the TSC today before merging

**Comment by mbolivar-nordic** [comment]
motion passed at TSC

**Pull request merged**
